crates/sui-indexer/src/handlers/checkpoint_handler.rs

```diff
@@ -156,143 +184,78 @@ where
         let mut next_cursor_sequence_number = last_seq_from_db + 1;
         // NOTE: we will download checkpoints in parallel, but we will commit them sequentially.
         // We will start with MAX_PARALLEL_DOWNLOADS, and adjust if no more checkpoints are available.
-        let mut current_parallel_downloads = MAX_PARALLEL_DOWNLOADS;
+        let current_parallel_downloads = MAX_PARALLEL_DOWNLOADS;
         loop {
-            let download_futures = (next_cursor_sequence_number
-                ..next_cursor_sequence_number + current_parallel_downloads as i64)
-                .map(|seq_num| self.download_checkpoint_data(seq_num as u64));
-            let download_results = join_all(download_futures).await;
-            let mut downloaded_checkpoints = vec![];
+            let mut download_futures = FuturesOrdered::new();
+            for seq_num in next_cursor_sequence_number
+                ..next_cursor_sequence_number + current_parallel_downloads as i64
+            {
+                download_futures.push_back(self.download_checkpoint_data(seq_num as u64));
+            }
             // NOTE: Push sequentially and if one of the downloads failed,
             // we will discard all following checkpoints and retry, to avoid messing up the DB commit order.
-            for download_result in download_results {
-                if let Ok(checkpoint) = download_result {
-                    downloaded_checkpoints.push(checkpoint);
-                } else {
-                    if let Err(IndexerError::UnexpectedFullnodeResponseError(fn_e)) =
-                        download_result
-                    {
-                        warn!(
-                            "Unexpected response from fullnode for checkpoints: {}",
-                            fn_e
-                        );
-                    } else if let Err(IndexerError::FullNodeReadingError(fn_e)) = download_result {
-                        warn!("Fullnode reading error for checkpoints {}: {}. It can be transient or due to rate limiting.", next_cursor_sequence_number, fn_e);
-                    } else {
-                        warn!("Error downloading checkpoints: {:?}", download_result);
+            while let Some(res) = download_futures.next().await {
```

Comment:
here we will download one by one, shall we keep the join_all and current_parallel_downloads adjustment logic?

Reply:
This does not download one by one. In the background tokio polls the `FuturesOrdered` and make them available as soon as the download has finished so we could `send` the checkpoint to the receiver. Thus this becomes the first pipeline in the process.

Before this change, we download X checkpoints at the same time and block wait until *ALL* of them have finished. When there is a checkpoint that takes long time (e.g. sent to a slow node), that will become the bottleneck of this step.